### PR TITLE
User: rich model

### PR DESCRIPTION
**Pull request by fprochazka**
[Depends #38] [Ref #30]

### CODE REVIEW

File: soccer-records-model/src/main/java/cz/muni/fi/pa165/hibernate/ProxyEntityManager.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    @PersistenceContext
+    private EntityManager entityManager;
+
+    /**
+     * CDI Producer, joins JTA transaction and returns this EntityManager.
+     */
+    @Produces
+    private EntityManager getEntityManager()
+    {
+        return (EntityManager) Proxy.newProxyInstance(
```
Comment (jkuchar) [inline soccer-records-model/src/main/java/cz/muni/fi/pa165/hibernate/ProxyEntityManager.java:31]: Little hard to read, maybe comment can fix this.

File: soccer-records-model/src/main/java/cz/muni/fi/pa165/user/User.java [hunk 2]
```diff
@@ -72,6 +73,10 @@ public String getPasswordHash()
      */
     public void promoteToAdmin()
     {
+        if (!this.role.allowTransition(UserRole.ADMIN)) {
```
Comment (jkuchar) [inline soccer-records-model/src/main/java/cz/muni/fi/pa165/user/User.java:76]: ❤️

File: soccer-records-model/src/main/java/cz/muni/fi/pa165/user/UserFacade.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+     */
+    public void promoteUserToModerator(UUID userId)
+    {
+        User user = userRepository.getUserById(userId);
+
+        userService.promoteUserToModerator(user);
+
+        entityManager.flush();
+    }
+
```
Comment (jkuchar) [inline soccer-records-model/src/main/java/cz/muni/fi/pa165/user/UserFacade.java:71]: 👍 Nice to test with BDD.

File: soccer-records-model/src/main/java/cz/muni/fi/pa165/user/UserRepositoryImpl.java [hunk 4]
```diff
@@ -55,9 +42,21 @@ public User findUserByEmail(final String email)
     }
 
     @Override
-    public User findUserById(UUID id)
+    public User getUserById(final UUID userId)
     {
-        return em.find(User.class, id);
+        if (userId == null) {
+            throw new IllegalArgumentException("Cannot search for a null id");
+        }
```
Comment (jkuchar) [inline soccer-records-model/src/main/java/cz/muni/fi/pa165/user/UserRepositoryImpl.java]: assert maybe? It's shorter. (Conceptually one should be validated by type system.)

File: soccer-records-model/src/main/java/cz/muni/fi/pa165/user/UserRepositoryImpl.java [hunk 5]
```diff
@@ -55,9 +42,21 @@ public User findUserByEmail(final String email)
...
+
+        try {
+            return entityManager
+                .createQuery("SELECT u FROM User u WHERE u.id = :userId", User.class)
+                .setParameter("userId", userId)
+                .getSingleResult();
+
+        } catch (NoResultException e) {
+            throw new UserNotFoundException(userId, e);
```
Comment (jkuchar) [inline soccer-records-model/src/main/java/cz/muni/fi/pa165/user/UserRepositoryImpl.java:56]: 👍 for not returning NULL

File: soccer-records-model/src/main/java/cz/muni/fi/pa165/user/UserService.java [hunk 6]
```diff
@@ -18,10 +19,33 @@ public UserService(PasswordEncoder passwordEncoder)
...
+        user.promoteToAdmin();
+    }
+
+    /**
+     * Promotes user to moderator.
+     */
+    public void promoteUserToModerator(User user)
+    {
+        user.promoteToModerator();
+    }
```
Comment (jkuchar) [inline soccer-records-model/src/main/java/cz/muni/fi/pa165/user/UserService.java:50]: I feel the testability all around! Are you? 😃

File: soccer-records-model/src/main/java/cz/muni/fi/pa165/user/exceptions/UserCannotBePromotedToRoleExceptions.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * @author Filip Prochazka <filip@prochazka.su>
+ */
+public class UserCannotBePromotedToRoleExceptions extends RuntimeException
+{
+
+    private final User user;
+    private final UserRole newRole;
+
+    public UserCannotBePromotedToRoleExceptions(final User user, final UserRole newRole)
```
Comment (jkuchar) [inline soccer-records-model/src/main/java/cz/muni/fi/pa165/user/exceptions/UserCannotBePromotedToRoleExceptions.java:15]: 👍 for formatting message inside exception.

File: soccer-records-model/src/main/resources/application-context.xml [hunk 8]
```diff
@@ -6,11 +6,13 @@
 
     <bean id="validator" class="org.springframework.validation.beanvalidation.LocalValidatorFactoryBean"/>
 
-    <bean id="userService" class="cz.muni.fi.pa165.user.UserService" autowire="constructor" />
-    <bean id="passwordEncoder" class="org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder">
+    <bean class="cz.muni.fi.pa165.user.UserFacade" autowire="constructor" />
+    <bean class="cz.muni.fi.pa165.user.UserService" autowire="constructor" />
+    <bean class="org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder">
```
Comment (jkuchar) [inline soccer-records-model/src/main/resources/application-context.xml:11]: alignment issue

File: soccer-records-model/src/test/java/cz/muni/fi/pa165/user/UserFacadeTest.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    public UserFacade userFacade;
+
+    @PersistenceContext
+    public EntityManager em;
+
+    @Test
+    public void testCreateUser() throws Exception
+    {
+        User createdUser = userFacade.createUser("filip@prochazka.su", "heslo");
+        em.clear();
```
Comment (jkuchar) [inline soccer-records-model/src/test/java/cz/muni/fi/pa165/user/UserFacadeTest.java:38]: 👍 for explicit flushes

File: soccer-records-model/src/test/java/cz/muni/fi/pa165/user/UserTest.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        assertEquals(UserRole.MODERATOR, user.getRole());
+    }
+
+    @Test(expectedExceptions = UserCannotBePromotedToRoleExceptions.class)
+    public void testCannotPromoteModeratorToModeratorException()
+    {
+        User user = getUserService().createUser("filip@prochazka.su", null, "heslo");
+
+        user.promoteToModerator();
+        user.promoteToModerator();
```
Comment (jkuchar) [inline soccer-records-model/src/test/java/cz/muni/fi/pa165/user/UserTest.java]: Why is this relation not reflective?

File: soccer-records-model/src/main/java/cz/muni/fi/pa165/user/UserFacade.java (hunk 3 above)
Comment (fprochazka) [reply soccer-records-model/src/main/java/cz/muni/fi/pa165/user/UserFacade.java:71]: I wasn't going to test this at all, and just test controllers directly... but hey, if we're not meeting the other requirements, let's at least write all the tests :)

File: soccer-records-model/src/main/java/cz/muni/fi/pa165/user/UserService.java (hunk 6 above)
Comment (fprochazka) [reply soccer-records-model/src/main/java/cz/muni/fi/pa165/user/UserService.java:50]: Love is in the air :heart:

File: soccer-records-model/src/main/resources/application-context.xml (hunk 8 above)
Comment (fprochazka) [reply soccer-records-model/src/main/resources/application-context.xml:11]: What do you mean?

File: soccer-records-model/src/test/java/cz/muni/fi/pa165/user/UserTest.java (hunk 10 above)
Comment (fprochazka) [reply soccer-records-model/src/test/java/cz/muni/fi/pa165/user/UserTest.java]: You think it should be?

**Pull request merged**